src/io/PerCellGenerator.cpp

```diff
@@ -0,0 +1,166 @@
+	if (_initTemperature > 0.) {
+		Log::global_log->info() << "initTemperature: " << _initTemperature << std::endl;
+	} else {
+		Log::global_log->error() << "Missing required field initTemperature. Aborting!" << std::endl;
+		Simulation::exit(1949);
+	}
+
+	xmlconfig.getNodeValue("generateAtLeastTwoParticles", _generateAtLeastTwoParticles);
+	Log::global_log->info() << "generateAtLeastTwoParticles: " << _generateAtLeastTwoParticles << std::endl;
+}
+
+void PerCellGenerator::fillContainer(ParticleContainer *particleContainer, Component *component,
+									 unsigned int numMoleculesPerCell, bool fillHalo) {
+	const auto cellLength = particleContainer->getCellLength();
+	const auto boxMin = std::array{particleContainer->getBoundingBoxMin(0), particleContainer->getBoundingBoxMin(1),
+								   particleContainer->getBoundingBoxMin(2)};
+	const auto boxMax = std::array{particleContainer->getBoundingBoxMax(0), particleContainer->getBoundingBoxMax(1),
+								   particleContainer->getBoundingBoxMax(2)};
+
+	std::array<unsigned int, 3> numCells{};
+	for (unsigned i = 0; i < 3; ++i) {
+		numCells[i] = static_cast<unsigned int>(std::round((boxMax[i] - boxMin[i]) / cellLength[i]));
+	}
+	auto haloWidthInNumCells = fillHalo ? particleContainer->getHaloWidthNumCells() : 0;
+#ifdef _OPENMP
+#pragma omp parallel
+#endif
+	{
+		std::random_device r;
+		std::default_random_engine randomEngine{r()};
+#ifdef _OPENMP
+#pragma omp for
+#endif
+		for (long xind = -haloWidthInNumCells; xind < numCells[0] + haloWidthInNumCells; ++xind) {
+			for (long yind = -haloWidthInNumCells; yind < numCells[1] + haloWidthInNumCells; ++yind) {
+				for (long zind = -haloWidthInNumCells; zind < numCells[2] + haloWidthInNumCells; ++zind) {
+					std::array<long, 3> cellInd3d{xind, yind, zind};
+					std::array<double, 3> cellMin{};
+					std::array<double, 3> cellMax{};
+					long startID =
+						xind * (numCells[1] + 2 * haloWidthInNumCells) * (numCells[2] + 2 * haloWidthInNumCells) +
+						yind * (numCells[2] + 2 * haloWidthInNumCells) + zind;
+					startID *= numMoleculesPerCell;
+					bool isHalo = false;
+					for (unsigned int i = 0; i < 3; ++i) {
+						if (cellInd3d[i] == -1) {
+							// lower halo
+							cellMin[i] = boxMin[i] + cellInd3d[i] * cellLength[i];
+							cellMax[i] = boxMin[i];
+							isHalo = true;
+						} else if (cellInd3d[i] == numCells[i] - 1) {
+							// upper non-halo
+							cellMin[i] = boxMin[i] + cellInd3d[i] * cellLength[i];
+							cellMax[i] = boxMax[i];
+						} else if (cellInd3d[i] == numCells[i]) {
+							// upper halo
+							cellMin[i] = boxMax[i];
+							cellMax[i] = boxMin[i] + cellInd3d[i] * cellLength[i];
+							isHalo = true;
+						} else {
+							// lower non-halo, or normal.
+							cellMin[i] = boxMin[i] + cellInd3d[i] * cellLength[i];
+							// cell position is calculated based on boxMin!
+							cellMax[i] = boxMin[i] + (cellInd3d[i] + 1) * cellLength[i];
+						}
```

Comment:
is this whole if/else tree necessary? E.g. the first case and the last else seem to be equivalent except for the `isHalo = true` in the end.

Reply:
yes, to handle the boundaries correctly and not insert halo particles by mistake.